### PR TITLE
Muon - Widget; Load Run Widget

### PR DESCRIPTION
**Pull request by TWJubb**
Refs #23642 

Requires #23670 to be merged.

This PR creates the load run widget of the new Muon Analysis interface.

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Muon/GUI/Common/load_run_widget/model.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+from __future__ import (absolute_import, division, print_function)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/model.py]: copyright needs adding

File: scripts/Muon/GUI/Common/load_run_widget/model.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    """Stores info on all currently loaded workspaces"""
+
+    def __init__(self, loaded_data_store=MuonLoadData()):
+        # Used with load thread
+        self._filenames = []
+
+        self._loaded_data_store = loaded_data_store
+        self._current_run = None
+
+    # TODO : remove the need for this function
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/model.py]: do we want this to do?

File: scripts/Muon/GUI/Common/load_run_widget/presenter.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        self._view.on_load_current_run_clicked(self.handle_load_current_run)
+        self._view.on_increment_run_clicked(self.handle_increment_run)
+        self._view.on_decrement_run_clicked(self.handle_decrement_run)
+        self._view.on_run_edit_changed(self.handle_run_changed_by_user)
+
+    def show(self):
+        self._view.show()
+
+    # used with threading
+    def create_load_thread(self):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/presenter.py]: this seems to be always assigned to a member of the class. We dont we do that here and then make it a void?

File: scripts/Muon/GUI/Common/load_run_widget/presenter.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        Updates list of runs by adding a run equal to 1 before to the lowest run.
+        """
+        run_list = self.flatten_run_list(copy.copy(self.runs))
+        if run_list is None or len(run_list) == 0:
+            return []
+        if len(run_list) == 1:
+            run_list = [run_utils.decrement_run(run_list[0]), run_list[0]]
+        else:
+            run_list = run_utils.decrement_run_list(run_list)
+        return run_list
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/presenter.py]: I have noticed that the start threading stuff has two very similar functions. Is there a way we could reduce them to a single function (with appropriate ifs)?

File: scripts/Muon/GUI/Common/load_run_widget/view.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        self.horizontal_layout.addWidget(self.load_current_run_button)
+        self.horizontal_layout.addWidget(self.decrement_run_button)
+        self.horizontal_layout.addWidget(self.instrument_label)
+        self.horizontal_layout.addWidget(self.run_edit)
+        self.horizontal_layout.addWidget(self.increment_run_button)
+
+        self.horizontal_layout.setContentsMargins(0, 0, 0, 0)
+        self.horizontal_layout.setMargin(0)
+
+    def set_item_sizes(self):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/view.py]: is this used in a later PR?

File: scripts/Muon/GUI/Common/load_run_widget/view.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        return str(self.instrument_label.text())
+
+    def set_instrument_label(self, text):
+        self.instrument_label.setText(text)
+
+    def set_current_instrument(self, instrument):
+        self.instrument_label.setText(instrument)
+
+    def set_run_edit_regex(self):
+        # "^[0-9]*([0-9]+[,-]{0,1})*[0-9]+$"
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/view.py]: I dont understand what the comment is suppose to signify

File: scripts/Muon/GUI/Common/load_run_widget/view.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+        regex = QtCore.QRegExp(run_utils.run_string_regex)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/load_run_widget/view.py]: should this be a utils function? Then we can test the RegExp

File: scripts/test/Muon/load_run_widget/loadrun_model_test.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    def test_model_is_cleared_correctly(self, load_utils_mock):
+        files = [r'EMU00019489.nxs', r'EMU00019490.nxs', r'EMU00019491.nxs']
+        load_return_vals = [([1, 2, 3], filename, 19489 + i) for i, filename in enumerate(files)]
+
+        model = LoadRunWidgetModel(MuonLoadData())
+        load_utils_mock.load_workspace_from_filename = mock.Mock()
+        load_utils_mock.load_workspace_from_filename.side_effect = load_return_vals
+
+        model.loadData(files)
+        model.execute()
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_model_test.py]: check it has loaded something first

File: scripts/test/Muon/load_run_widget/loadrun_presenter_current_run_test.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+if sys.version_info.major == 3:
+    from unittest import mock
+else:
+    import mock
+
+from PyQt4 import QtGui
+from PyQt4.QtGui import QApplication
+
+# global QApplication (get errors if > 1 instance in the code)
+QT_APP = QApplication([])
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_presenter_current_run_test.py]: I think for this you could use https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/mock_widget.py as seen in https://github.com/mantidproject/mantid/blob/master/scripts/test/Muon/transformWidget_test.py#L27. Unless I have missed something

File: scripts/Muon/GUI/Common/load_run_widget/presenter.py (hunk 3 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/load_run_widget/presenter.py]: I think the code is clearer in it's current form and we don't really gain anything by assigning within this function?

Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/load_run_widget/presenter.py]: I cant type today. I just wanted to know why it was done this way

File: scripts/Muon/GUI/Common/load_run_widget/presenter.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+            )
+            # remove the single loaded filename
+            self._model._loaded_data_store.remove_data(run=run)
+        self._model._loaded_data_store.remove_data(run=run_list[0])
+        # run_string = runUtils.run_list_to_string(run_list)
+        return_ws["OutputWorkspace"] = MuonWorkspaceWrapper(running_total)
+        self._model._loaded_data_store.add_data(run=self.flatten_run_list(run_list), workspace=return_ws,
+                                                filename="Co-added")
+
+    def empty(self):
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/load_run_widget/presenter.py]: As the answer is as far as I can tell nothing I have removed it

File: scripts/test/Muon/load_run_widget/loadrun_presenter_multiple_file_test.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+from PyQt4 import QtGui
+from PyQt4.QtGui import QApplication
+
+if sys.version_info.major == 3:
+    from unittest import mock
+else:
+    import mock
+
+# global QApplication (get errors if > 1 instance in the code)
+QT_APP = QApplication([])
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_presenter_multiple_file_test.py]: I think we have code to do this already

File: scripts/test/Muon/load_run_widget/loadrun_presenter_multiple_file_test.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+        six.assertCountEqual(self, self.model.loaded_filenames, ["file2.nxs", "file3.nxs", "file4.nxs", "file5.nxs"])
+        six.assertCountEqual(self, self.model.loaded_workspaces, [[2], [3], [4], [5]])
+        six.assertCountEqual(self, self.model.loaded_runs, [2, 3, 4, 5])
+
+        self.assertEqual(self.view.get_run_edit_text(), "2-5")
+
+    @run_test_with_and_without_threading
+    def test_that_if_decrement_run_fails_the_data_are_returned_to_previous_state(self):
+        def load_failure():
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_presenter_multiple_file_test.py]: could we move this as it is used in a couple of places

File: scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+
+        self.presenter.handle_run_changed_by_user()
+        self.wait_for_thread(self.presenter._load_thread)
+
+        self.assertEqual(self.presenter.filenames, ["1234.nxs"])
+        self.assertEqual(self.presenter.runs, [1234])
+        self.assertEqual(self.presenter.workspaces, [[1]])
+
+
+class LoadRunWidgetIncrementDecrementSingleFileModeTest(unittest.TestCase):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py]: should this be in a different file?

File: scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        self.model.load_workspace_from_filename = mock.Mock(side_effect=self.load_failure)
+
+    # ------------------------------------------------------------------------------------------------------------------
+    # TESTS : Test the increment/decrement buttons in single file mode (can only load one run at a time)
+    # ------------------------------------------------------------------------------------------------------------------
+
+    @run_test_with_and_without_threading
+    def test_that_decrement_run_attempts_to_load_the_correct_run(self):
+        new_filename = "EMU00001233.nxs"
+        load_call_count = self.load_utils_patcher.load_workspace_from_filename.call_count
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py]: should this copy the values explicitly? As it seems to be part of comparison below

File: scripts/Muon/GUI/Common/load_run_widget/view.py (hunk 6 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/load_run_widget/view.py]: I think the comment is the regex string this applies i.e. the contents of `run_utils.run_string_regex`

File: scripts/Muon/GUI/Common/load_run_widget/view.py (hunk 7 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/load_run_widget/view.py]: I think the argument for it being in the view is that is uses Qt functionality which we are trying to keep confined to the view. Having said that I think that testing the regex is a good idea I'll add some tests to the view

File: scripts/Muon/GUI/Common/load_run_widget/view.py (hunk 6 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/load_run_widget/view.py]: can we just say that?

File: scripts/Muon/GUI/Common/load_run_widget/view.py (hunk 7 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/load_run_widget/view.py]: ok

File: scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py (hunk 14 above)
Comment (Mantid-Matthew) [reply scripts/test/Muon/load_run_widget/loadrun_presenter_single_file_test.py]: new_filename is I think the expected value of the filename after it has been decremented. At the start of the test the filename is `EMU00001234.nxs`

File: scripts/Muon/GUI/Common/load_run_widget/model.py (hunk 2 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/load_run_widget/model.py]: I can't quite work out why it is isn't needed currently.

File: scripts/test/Muon/load_run_widget/loadrun_view_test.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class LoadRunWidgetViewTest(unittest.TestCase):
+
+    def setUp(self):
+        self._qapp = mock_widget.mockQapp()
+        self.view = LoadRunWidgetView()
+        self.view.run_edit.insert('')
+
+    def test_run_number_line_edit_doesnt_accept_alphabetical(self):
```
Comment (AnthonyLim23) [inline scripts/test/Muon/load_run_widget/loadrun_view_test.py]: can we check that a mix of number and letters fails? e.g. 22-3a?

**Review by AnthonyLim23 — APPROVED** [review]
looks good

**Comment by AndreiSavici** [comment]
Is this still in progress? The required ticket was merged

**Pull request merged**